src/c_api.cpp

```diff
@@ -173,6 +173,46 @@ class Booster {
     }
   }
 
+  static void CheckDatasetResetConfig(const Config& old_config, const std::unordered_map<std::string, std::string>& new_param) {
+    Config new_config;
+    new_config.Set(new_param);
+    if (new_param.count("max_bin") && new_config.max_bin != old_config.max_bin) {
+      Log::Warning("Cannot change max_bin after constructed Dataset handle.");
```

Comment:
Not sure, but maybe `Fatal` is better here?.. I think it will help to avoid any confusion about why changing params didn't take effect. I mean, it seems to me that the most users either set `verbose=-1`, or use Jupyter Notebook where LightGBM's log is printed in Jupyter console, and in consequence simply cannot see any warning.